File: spec/01_arrays_and_strings/1_4_palindrome_permutation.rb
# n - str length, a - str alphabet size

# B: time = O(n), mem = O(a)
def palindrome_permutation_b?(str)
  h = Hash.new { |hash, key| hash[key] = 0 }

  str.each_char do |c|
    if c != ' '
      h[c] += 1
    end
  end

  odd_pairs = 0

  h.each do |_, c|
    if c % 2 > 0
      odd_pairs += 1
    end
  end

  odd_pairs <= 1
end

# C: time = O(n * lon(n)), mem = O(1)
def palindrome_permutation_c?(str)
  str = sort!(str)

  odd_pairs = 0
  last_char = str[0]
  last_char_count = 1
  i = 1

  while i < str.size do
    if str[i] == last_char
      last_char_count += 1
    else

      if last_char_count % 2 > 0 && last_char != ' '
        odd_pairs += 1
      end

      last_char = str[i]
      last_char_count = 1
    end

    i += 1
  end

  if last_char_count % 2 > 0 && last_char != ' '
    odd_pairs += 1
  end

  odd_pairs <= 1
end

RSpec.describe 'palindrome_permutation' do
  %i(b c).each do |implementation|
    describe "#{implementation} case" do
      subject { send("palindrome_permutation_#{implementation}?", str) }

      context 'palindrome permutation' do
        let(:str) { 'taco cat' }

        it { is_expected.to eq true }
      end

      context 'not a palindrome permutation' do
        let(:str) { 'taco cata' }

        it { is_expected.to eq false }
      end
    end
  end
end

File: go/01_arrays_and_strings/1_4_palindrome_permutation.go
package arrays

// n - str length, a - str alphabet size
// B: time = O(n), mem = O(a)
func PalindromePermutationB(str string) bool {
	h := map[rune]int{}

	for _, c := range str {
		if c != ' ' {
			if _, ok := h[c]; ok {
				h[c] += 1
			} else {
				h[c] = 1
			}
		}
	}

	oddPairs := 0

	for _, c := range h {
		if c%2 > 0 {
			oddPairs += 1
		}
	}

	return oddPairs <= 1
}

// C: time = O(n * lon(n)), mem = O(1)
func PalindromePermutationC(str string) bool {
	str = SortStringByCharacter(str)

	oddPairs := 0
	lastChar := str[0]
	lastCharCount := 1

	for i := 1; i < len(str); i++ {
		if str[i] == lastChar {
			lastCharCount += 1
		} else {
			if lastCharCount%2 > 0 && lastChar != ' ' {
				oddPairs += 1
			}

			lastChar = str[i]
			lastCharCount = 1
		}
	}

	if lastCharCount%2 > 0 && lastChar != ' ' {
		oddPairs += 1
	}

	return oddPairs <= 1
}
